### ISSUE TITLE
Mosquitto as its own webserver not working

### ISSUE THREAD

**Issue by cymplecy** [eclipse/mosquitto#286]
My interpretation of the docs, makes me think that mosquitto should work as its own simple webserver for websockets use (To save having to install a separate one just to host a web page for processing mqtt javascript based webpage) 

However, I cannot get this to work (On a RaspberryPi using 1.4.10)

I am using this in my conf file
listener 80
protocol websockets
http_dir /var/www/html

If I run lighttpd serving the content from /var/www/html instead of the above it all works fine but it won;t work without it.

Have I mis-interpreted this bit of mosquito capability, got the wrong configuration or does it not work?

**Comment by ralight** [comment]
If you're on Raspbian Jessie then the version of libwebsockets available does not provide good web support, you'll have to install from scratch I'm afraid. You should go for libwebsockets version 1.7.x or 2.x.

**Comment by cymplecy** [comment]
Thanks for info.  I'll live with using lighttpd for now till Jessie catches up

**Comment by ssimd** [comment]
I ran into the same problem: the HTTP connection was always reset -> ERR_EMPTY_RESPONSE in Chrome. Looking at the packet traffice, there were only some TCP packets transferred, of which FIN ACK is sent quite immediately, thus terminating the connection.

However, I figured out the problem exists in recent version 1.4.10 of mosquitto.

But when I use the version 1.4.8 from debian repositories, it works fine. The same for version 1.4.8 on an Ubuntu.

**Comment by git-oliver** [comment]
I built from source: mosquitto v1.4.8, v1.4.9, v1.4.10, master and develop branch. None of them work correctly as web-server. I also tried it by upgrading to libwebsockets8 2.0.3-2 (including newest *-dev and all deps) - no success.

@ssimd: AFAIK 1.4.8 comes without websockets support, isn't it? Or did you find a solution with websockets support that also works as a web-server?

**Comment by ralight** [comment]
test.mosquitto.org uses libwebsockets 2.0.2 compiled with default options. mosquitto is 1.4.10 compiled with websockets support and the relevant part of the config file is

```
listener 8080
protocol websockets
http_dir /var/lib/mosquitto/http_dir
```

That works fine. I don't know what could be your problem.

**Comment by ssimd** [comment]
mosquitto 1.4.8 in the debian stretch repository is with websocket support.
currently I have a working setup in a docker container with the following versions:
mosquitto:amd64  1.4.8-1+b1
libwebsockets7:amd64  1.7.3-1

however, I built this container on Nov 15, since then the version was updated in the repo to 1.4.10

my not-working setup uses mosquitto 1.4.10-0mosquitto4 and libwebsockets8 Version: 2.0.3-2

need to rebuilt the container and see how it looks / play around with some versions.

**Comment by git-oliver** [comment]
I was wrong: mosquitto 1.4.8 does support websockets and also works as webserver!

Meanwhile I found:
* Mint 18.x has mosquitto 1.4.8 and libwebsockets7 1.7.1 and it works as webserver.
* Raspbian Jessie has mosquitto 1.3.4 without websockets/webserver support.
* mosquitto.org has a repo for newer Raspbian mosquitto 1.4.10. But together with Raspbians libwebsockets3 1.2.2 it does not work as webserver
* On Raspbian I built from source the latest libwebsockets (2.1.0) and mosquitto (1.4.10, branch "fixes"!) and now mosquitto works as webserver and with websockets support  :-)

IMHO this bug is caused by libwebsockets and not mosquitto alone.
